### PR TITLE
overlord/snapstate: make "snap remove" remove all revisions of a snap

### PR DESCRIPTION
**Pull request by pedronis**
This also adds more tracking of foreign tasks in the snapstate tests.

### CODE REVIEW

**Comment by pedronis** [comment]
requires #1063

**Comment by pedronis** [comment]
retest this please

**Comment by chipaca** [comment]
can you rebase please?

**Comment by pedronis** [comment]
@chipaca done

**Comment by chipaca** [comment]
:+1:

File: overlord/snapstate/export_test.go [hunk 1]
```diff
@@ -38,10 +38,27 @@ func SetSnapstateBackend(b ManagerBackend) {
...
+type ForeignTaskTracker interface {
+	ForeignTask(kind string, status state.Status, ss *SnapSetup)
+}
+
 // AddForeignTaskHandlers registers handlers for tasks handled outside of the snap manager.
-func (m *SnapManager) AddForeignTaskHandlers() {
+func (m *SnapManager) AddForeignTaskHandlers(tracker ForeignTaskTracker) {
 	// Add fake handlers for tasks handled by interfaces manager
-	fakeHandler := func(task *state.Task, _ *tomb.Tomb) error { return nil }
+	fakeHandler := func(task *state.Task, _ *tomb.Tomb) error {
```
Comment (zyga) [inline overlord/snapstate/export_test.go]: Nice!

File: overlord/snapstate/snapmgr_test.go [hunk 2]
```diff
@@ -862,10 +866,14 @@ version: 1.0`)
 	c.Check(s.fakeBackend.ops[3].name, Equals, "/snap/mock/100003")
 	c.Check(s.fakeBackend.ops[3].old, Equals, "/snap/mock/100002")
 
-	c.Check(s.fakeBackend.ops[4].op, Equals, "candidate")
-	c.Check(s.fakeBackend.ops[4].sinfo, DeepEquals, snap.SideInfo{Revision: 100003})
-	c.Check(s.fakeBackend.ops[5].op, Equals, "link-snap")
-	c.Check(s.fakeBackend.ops[5].name, Equals, "/snap/mock/100003")
+	c.Check(s.fakeBackend.ops[4].op, Equals, "setup-profiles:Doing")
+	c.Check(s.fakeBackend.ops[4].name, Equals, "mock")
+	c.Check(s.fakeBackend.ops[4].revno, Equals, 100003)
```
Comment (zyga) [inline overlord/snapstate/snapmgr_test.go]: Very nice indeed!

File: overlord/snapstate/snapmgr_test.go [hunk 3]
```diff
@@ -930,28 +938,159 @@ func (s *snapmgrTestSuite) TestRemoveIntegration(c *C) {
...
+			op:     "can-remove",
+			name:   "/snap/some-snap/7",
+			active: true,
+		},
+		{
+			op:   "unlink-snap",
+			name: "/snap/some-snap/7",
+		},
+		{
+			op:    "remove-profiles:Doing",
```
Comment (zyga) [inline overlord/snapstate/snapmgr_test.go]: Where do we stop services for the affected snap?

**Comment by zyga** [comment]
+1

File: overlord/snapstate/snapmgr_test.go (hunk 3 above)
Comment (pedronis) [inline overlord/snapstate/snapmgr_test.go]: unlink-snap calls snappy.UnlinkSnap which calls removePackageServices (which afaict is also responsible for first stopping them)

File: overlord/snapstate/snapmgr_test.go [hunk 4]
```diff
@@ -930,28 +938,159 @@ func (s *snapmgrTestSuite) TestRemoveIntegration(c *C) {
...
+		c.Check(ss, DeepEquals, expSnapSetup, Commentf(t.Kind()))
+	}
+
+	// verify snaps in the system state
+	var snapst snapstate.SnapState
+	err = snapstate.Get(s.state, "some-snap", &snapst)
+	c.Assert(err, Equals, state.ErrNoState)
+}
+
+func (s *snapmgrTestSuite) TestRemoveWithManyRevisionsIntegration(c *C) {
```
Comment (elopio) [inline overlord/snapstate/snapmgr_test.go]: :+1: this is good.
Do you think an additional user test would be useful here? One that calls snap install twice, snap remove once, and verifies that snap list doesn't show the snap.

**Comment by mvo5** [comment]
Looks very nice! Thank you.

**Pull request merged**